### PR TITLE
ABPI-83846

### PR DESCRIPTION
**Pull request by mmiesbach**
- Using Androids bluetooth apis to detect connected bluetooth devices
- Showing those devices in a list
- If there are no connected devices, the list shows "no connected devices"
- The list automatically updates with new data while the screen is visible

### CODE REVIEW

File: app/src/main/java/com/example/bluetoothdevicesapplication/ui/view/DevicesFragment.kt [hunk 1]
```diff
@@ -1,38 +1,71 @@
...
-
-        return view
+        return inflater.inflate(R.layout.devices_fragment, container, false)
     }
 
     override fun onActivityCreated(savedInstanceState: Bundle?) {
         super.onActivityCreated(savedInstanceState)
         viewModel = ViewModelProvider(this).get(DevicesViewModel::class.java)
-        // TODO: Use the ViewModel
+        viewModel.pairedDevices.observe(viewLifecycleOwner, {
```
Comment (flowolf86) [inline app/src/main/java/com/example/bluetoothdevicesapplication/ui/view/DevicesFragment.kt:44]: Very good, the lifecycle concept in Android is important, helps you prevent leaks and keeps your app performant

File: app/src/main/java/com/example/bluetoothdevicesapplication/ui/view/DevicesFragment.kt [hunk 2]
```diff
@@ -1,38 +1,71 @@
...
+            recyclerView.layoutManager = LinearLayoutManager(requireContext())
+            recyclerView.adapter = DevicesRecyclerViewAdapter(it)
+            noDevicesTextView.isVisible = it.isEmpty()
+        })
+
+        broadcastReceiver = object : BroadcastReceiver() {
+            override fun onReceive(p0: Context?, p1: Intent?) {
+                    viewModel.getPairedDevices()
+                    recyclerView.adapter?.notifyDataSetChanged()
```
Comment (flowolf86) [inline app/src/main/java/com/example/bluetoothdevicesapplication/ui/view/DevicesFragment.kt]: you can think about only updating the list if your data changed, but that also works

File: app/src/main/java/com/example/bluetoothdevicesapplication/ui/view/DevicesFragment.kt [hunk 3]
```diff
@@ -1,38 +1,71 @@
...
+            }
+        }
+
+    }
+
+    override fun onResume() {
+        super.onResume()
+        viewModel.getPairedDevices()
```
Comment (flowolf86) [inline app/src/main/java/com/example/bluetoothdevicesapplication/ui/view/DevicesFragment.kt]: Do you really GET the devices here, or just tell the viewmodel to refresh them. Think about the method naming

File: app/src/main/java/com/example/bluetoothdevicesapplication/ui/view/DevicesRecyclerViewAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DevicesRecyclerViewHolder {
+        val view = LayoutInflater.from(parent.context).inflate(R.layout.list_item ,parent, false)
+
+        return DevicesRecyclerViewHolder(view)
+    }
+
+    override fun onBindViewHolder(holder: DevicesRecyclerViewHolder, position: Int) {
+        val currentDevice = pairedDevices[position]
+        holder.textView.text = currentDevice.name
+        val m: Method = currentDevice.javaClass.getMethod("isConnected")
```
Comment (flowolf86) [inline app/src/main/java/com/example/bluetoothdevicesapplication/ui/view/DevicesRecyclerViewAdapter.kt:27]: Is there really no other way do do this than reflection? Well done finding this, Maybe this helps:

https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/android/bluetooth/BluetoothManager.java;l=140?q=BluetoothManager

File: app/src/main/java/com/example/bluetoothdevicesapplication/ui/viewmodel/DevicesViewModel.kt [hunk 5]
```diff
@@ -1,7 +1,27 @@
...
+import android.bluetooth.BluetoothDevice
+import android.bluetooth.BluetoothManager
+import android.content.Context
+import androidx.lifecycle.AndroidViewModel
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+
+class DevicesViewModel(application: Application) : AndroidViewModel(application) {
+    private val _pairedDevices = MutableLiveData<List<BluetoothDevice>>()
+    val pairedDevices : LiveData<List<BluetoothDevice>>
```
Comment (flowolf86) [inline app/src/main/java/com/example/bluetoothdevicesapplication/ui/viewmodel/DevicesViewModel.kt:13]: Well done not exposing the MutableLiveData!

**Review by flowolf86 — CHANGES_REQUESTED** [review]
This looks very well written already, good job! Nice implemention of architecture principles, use of APIs and clean code. I just have a few comments

File: app/src/main/java/com/example/bluetoothdevicesapplication/ui/view/DevicesFragment.kt (hunk 2 above)
Comment (mmiesbach) [reply app/src/main/java/com/example/bluetoothdevicesapplication/ui/view/DevicesFragment.kt]: Maybe I can let getPairedDevices return true if data was changed, false if not and then notify only if it returns true, would that be a fine solution?

File: app/src/main/java/com/example/bluetoothdevicesapplication/ui/view/DevicesFragment.kt (hunk 3 above)
Comment (mmiesbach) [reply app/src/main/java/com/example/bluetoothdevicesapplication/ui/view/DevicesFragment.kt]: Ah you are right ^^, I guess I will change it to "updatePairedDevices"

File: app/src/main/java/com/example/bluetoothdevicesapplication/ui/view/DevicesRecyclerViewAdapter.kt (hunk 4 above)
Comment (mmiesbach) [reply app/src/main/java/com/example/bluetoothdevicesapplication/ui/view/DevicesRecyclerViewAdapter.kt:27]: This sadly didn't work for me, it always returned an empty list. I did some research and it seems that it only works for GATT devices (has something to do with Bluetooth low energy devices) [https://stackoverflow.com/a/45327459](https://stackoverflow.com/a/45327459). 
I got this solution from [https://stackoverflow.com/a/58882930](https://stackoverflow.com/a/58882930) as nothing else worked for me but maybe I am missing something.

File: app/src/main/java/com/example/bluetoothdevicesapplication/ui/view/DevicesFragment.kt (hunk 2 above)
Comment (flowolf86) [reply app/src/main/java/com/example/bluetoothdevicesapplication/ui/view/DevicesFragment.kt]: Yes, that's acceptable

File: app/src/main/java/com/example/bluetoothdevicesapplication/ui/view/DevicesRecyclerViewAdapter.kt (hunk 4 above)
Comment (flowolf86) [reply app/src/main/java/com/example/bluetoothdevicesapplication/ui/view/DevicesRecyclerViewAdapter.kt:27]: I see, yeah might be that we are using internal AOSP APIs for that, or you could try the profile itself, but that goes too far. Reflection it is.

**Pull request merged**